### PR TITLE
Fix review app status URL normalization

### PR DESCRIPTION
**Pull request by justin808**
## Summary

- Derive review-app status links from the deployed GVC `app_domain` when available.
- Keep the raw Control Plane workload endpoint as the fallback when the endpoint cannot be parsed or matched to the expected `.cpln.app` host format.
- Support Control Plane org-prefixed endpoint hosts by stripping the org endpoint suffix from the derived GVC alias.
- Preserve path/query app-domain templates instead of blindly appending a trailing slash.
- Add generator coverage and a behavioral spec that executes the embedded URL-derivation Ruby against representative endpoints.

## Why

Downstream apps should not need local post-processing jobs to rewrite the review-app URL in deployment statuses and comments. Centralizing this in the reusable workflow also avoids fragile downstream `gh api -f` body updates and deployment-selection fallbacks.

## Verification

- `CPLN_ORG=dummy bundle exec rspec spec/command/generate_github_actions_spec.rb`
- `bundle exec rubocop spec/command/generate_github_actions_spec.rb`
- `ruby -e "require 'yaml'; Dir['.github/workflows/*.yml', '.github/actions/**/action.yml'].sort.each { |path| YAML.safe_load_file(path, aliases: true) }; puts 'yaml ok'"`
- `actionlint -ignore 'property "workflow_(ref|sha|repository|file_path)" is not defined in object type' .github/workflows/cpflow-deploy-review-app.yml`
- `.agents/bin/docs`

`CPLN_ORG=dummy .agents/bin/validate` was attempted, but the full suite exercises live Control Plane integration paths and failed because org `dummy` does not exist. The focused generator spec above covers this workflow change.

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected review-app status links to use the deployed application domain when available.
  * Added fallback handling so links continue to work when domain configuration is unavailable or invalid.
* **Tests**
  * Added coverage for domain-based URL generation and edge cases.
* **Documentation**
  * Updated the unreleased changelog with the corrected review-app link behavior.

### CODE REVIEW

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 1]
```diff
@@ -463,15 +463,52 @@ jobs:
...
+            gvc_json="$(cpln gvc get "${APP_NAME}" --org "${CPLN_ORG}" -o json 2>/dev/null || true)"
+            app_domain_template="$(
+              jq -r '.spec.env[]? | select(.name == "app_domain") | .value // empty' <<< "${gvc_json}" | head -n 1
+            )"
+
+            if [[ -n "${app_domain_template}" ]]; then
+              app_url="$(
+                ruby -ruri -e '
+                  workload_name, workload_url, app_domain_template, app_name = ARGV
+                  uri = URI(workload_url)
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml]: Fixed in `0041295`: the embedded Ruby now rescues `URI::InvalidURIError` and prints the raw workload endpoint, so malformed endpoint values fall back instead of aborting before outputs are written. Added behavioral coverage for that fallback.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 2]
```diff
@@ -463,15 +463,52 @@ jobs:
...
+                  host_prefix = "#{workload_name}-"
+                  host_suffix = ".cpln.app"
+
+                  unless uri.host&.start_with?(host_prefix) && uri.host&.end_with?(host_suffix)
+                    puts workload_url
+                    exit
+                  end
+
+                  cpln_gvc_alias = uri.host.delete_prefix(host_prefix).delete_suffix(host_suffix)
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml]: Fixed in `0041295`: after stripping the workload prefix and `.cpln.app`, the code now takes the first label as the GVC alias, so `rails-abc123.org-prefix.cpln.app` resolves to `abc123`. Added a behavioral spec for that org-prefixed endpoint shape.

File: .github/workflows/cpflow-deploy-review-app.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml]: Fixed in `0041295`: `URI(workload_url)` is wrapped in a rescue that falls back to the raw endpoint, matching the PR fallback goal. The focused spec now exercises malformed endpoint fallback.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 3]
```diff
@@ -463,15 +463,52 @@ jobs:
...
+                  cpln_gvc_alias_token = "$" + "(CPLN_GVC_ALIAS)"
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml:507]: Fixed in `0041295`: added an inline comment explaining that the concatenated token avoids actionlint flagging a literal command-substitution token inside single-quoted Ruby.

File: spec/command/generate_github_actions_spec.rb [hunk 4]
```diff
@@ -656,6 +656,21 @@ def shared_workflow_path(name)
       expect(contents).to include("curl_max_time: ${{ vars.REVIEW_APP_HEALTH_CHECK_CURL_MAX_TIME || '10' }}")
     end
 
+    it "prefers the deployed app_domain for review-app links when available" do
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/command/generate_github_actions_spec.rb:677]: Fixed in `0041295`: added a behavioral spec that extracts and executes the embedded Ruby for simple `.cpln.app` endpoints, org-prefixed endpoints, malformed endpoints, and non-cpln custom endpoints.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 5]
```diff
@@ -463,15 +463,52 @@ jobs:
           set -euo pipefail
           workload_name="${PRIMARY_WORKLOAD:-rails}"
           workload_url="$(cpln workload get "${workload_name}" --gvc "${APP_NAME}" --org "${CPLN_ORG}" -o json | jq -r '.status.endpoint // empty')"
-          echo "workload_url=${workload_url}" >> "$GITHUB_OUTPUT"
+          app_url="${workload_url}"
+
+          if [[ -n "${workload_url}" ]]; then
+            gvc_json="$(cpln gvc get "${APP_NAME}" --org "${CPLN_ORG}" -o json 2>/dev/null || true)"
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml]: Fixed in `0041295`: the GVC lookup now captures stderr, emits a GitHub warning, prints the cpln error to stderr, and then falls back to the raw workload endpoint.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 6]
```diff
@@ -463,15 +463,52 @@ jobs:
...
+                    exit
+                  end
+
+                  cpln_gvc_alias = uri.host.delete_prefix(host_prefix).delete_suffix(host_suffix)
+                  cpln_gvc_alias_token = "$" + "(CPLN_GVC_ALIAS)"
+                  app_url = app_domain_template
+                    .gsub(cpln_gvc_alias_token, cpln_gvc_alias)
+                    .gsub("{{APP_NAME}}", app_name)
+
+                  app_url = "#{app_url}/" unless app_url.end_with?("/")
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml]: Fixed in `0041295`: trailing slash normalization now applies only when the substituted app URL has no path, query, or fragment. The behavioral spec covers a path/query app-domain template and verifies no slash is appended.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 7]
```diff
@@ -463,15 +463,52 @@ jobs:
...
+                  puts app_url
+                ' "${workload_name}" "${workload_url}" "${app_domain_template}" "${APP_NAME}"
+              )"
+            fi
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml:524]: Fixed in `0041295`: this is covered by the same `URI::InvalidURIError` rescue and malformed-endpoint behavioral spec, so the raw endpoint remains the fallback instead of aborting the step.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 8]
```diff
@@ -463,15 +463,75 @@ jobs:
...
+                  host_prefix = "#{workload_name}-"
+                  host_suffix = ".cpln.app"
+
+                  unless uri.host&.start_with?(host_prefix) && uri.host&.end_with?(host_suffix)
+                    puts workload_url
+                    exit
+                  end
+
+                  raw_cpln_gvc_alias = uri.host.delete_prefix(host_prefix).delete_suffix(host_suffix)
+                  cpln_gvc_alias = raw_cpln_gvc_alias.split(".").first
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml:501]: Fixed in commit 61d2840. Empty aliases now fall back to the raw workload URL before template substitution, and the regression case is covered by the focused spec.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 9]
```diff
@@ -463,15 +463,75 @@ jobs:
...
+            gvc_stderr="$(mktemp)"
+            if ! gvc_json="$(cpln gvc get "${APP_NAME}" --org "${CPLN_ORG}" -o json 2>"${gvc_stderr}")"; then
+              echo "::warning::Could not read GVC app_domain; falling back to workload endpoint."
+              sed 's/^/cpln gvc get: /' "${gvc_stderr}" >&2
+              gvc_json=""
+            fi
+            rm -f "${gvc_stderr}"
+
+            app_domain_template="$(
+              jq -r '.spec.env[]? | select(.name == "app_domain") | .value // empty' <<< "${gvc_json}" | head -n 1
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml:478]: [auto-deferred]

This documentation expansion is outside the focused workflow-fix paths and would restart the final review cycle. The behavior remains opt-in and safely falls back to the workload endpoint when app_domain is absent; the user-visible change is recorded in CHANGELOG.md.

File: spec/command/generate_github_actions_spec.rb [hunk 10]
```diff
@@ -656,6 +674,60 @@ def shared_workflow_path(name)
...
+        )
+      ).to eq("not a uri")
+
+      expect(
+        run_review_app_url_script(
+          workload_name: "rails",
+          workload_url: "https://custom.example.test",
+          app_domain_template: "https://rails-$(CPLN_GVC_ALIAS).example.test",
+          app_name: "hichee-review-1"
+        )
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/command/generate_github_actions_spec.rb:736]: The empty-alias regression is fixed and covered in commit 61d2840. Focused specs now exercise that crash path, with actionlint and YAML validation covering the workflow structure.

[auto-deferred]

A full bash-level cpln/jq mocking harness is broader than this focused fix; the existing fallback branches remain structurally asserted and the direct Ruby behavior is exercised end to end.

**Comment by justin808** [comment]
Workflow Change Audit:

- Semantic classification: semantic reusable-workflow behavior change. The deployment-status URL now prefers the configured app domain and preserves the workload endpoint as fallback.
- Changed workflow/action files: `.github/workflows/cpflow-deploy-review-app.yml`.
- Triggers: unchanged.
- Permissions: unchanged.
- Secret references: unchanged.
- Third-party actions: none added or version-changed.
- Command parsing/execution: embedded Ruby URL normalization now rejects an empty derived alias by returning the original workload endpoint; regression coverage is in commit `61d2840`.
- Stale-base race control: merged current `main` at `6780910`, regenerated command docs for the 5.2.0 base at `7a191f6`, and will require current-head CI before merge.
- Local validation: focused generator spec (68 examples, 0 failures), RuboCop clean, generated command docs clean, YAML parse clean, and actionlint clean with the repository's known unsupported `job.workflow_*` diagnostics ignored.
- Required post-merge exercise: https://github.com/shakacode/control-plane-flow/issues/407

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 11]
```diff
@@ -463,15 +463,79 @@ jobs:
...
+                    uri = URI(workload_url)
+                  rescue URI::InvalidURIError
+                    puts workload_url
+                    exit
+                  end
+
+                  host_prefix = "#{workload_name}-"
+                  host_suffix = ".cpln.app"
+
+                  unless uri.host&.start_with?(host_prefix) && uri.host&.end_with?(host_suffix)
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml:495]: [auto-deferred]

This is a readability-only nit after the final candidate. The existing nil-safe guard and subsequent validated host use are correct, so another push would add review/CI churn without changing behavior.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 12]
```diff
@@ -463,15 +463,79 @@ jobs:
...
+                  puts app_url
+                ' "${workload_name}" "${workload_url}" "${app_domain_template}" "${APP_NAME}"
+              )"
+            fi
+          fi
+
+          {
+            echo "workload_url=${workload_url}"
+            echo "app_url=${app_url}"
+          } >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml:530]: [auto-deferred]

Defense-in-depth output hardening would broaden this focused URL fix and the pre-existing workload_url path. The value is controlled by trusted org configuration rather than PR input; current validation covers the intended templates and fallbacks, so this is not worth restarting the final review cycle.

File: spec/command/generate_github_actions_spec.rb [hunk 13]
```diff
@@ -58,6 +59,23 @@ def reusable_pr_open_help_workflow_path
     shared_workflow_path("cpflow-review-app-help")
   end
 
+  def review_app_url_ruby_script
+    match = reusable_review_app_workflow_path.read.match(
+      /ruby -ruri -e '\n(?<script>.*?)\n\s*' "\$\{workload_name\}"/m
+    )
+
+    match[:script].lines.map { |line| line.delete_prefix("                  ") }.join
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/command/generate_github_actions_spec.rb:67]: [auto-deferred]

This would improve a future test failure message but does not change production behavior or coverage. The extraction is currently exercised by 68 passing focused examples; changing it now would be a nit-only final-candidate push.

File: CHANGELOG.md [hunk 14]
```diff
@@ -32,6 +32,10 @@ In addition to the standard keepachangelog.com categories, this project uses a l
 - **Fixed generated review-app deploy workflows so they wait for workload readiness and an accepted HTTP response before marking a GitHub deployment successful.** [PR 363](https://github.com/shakacode/control-plane-flow/pull/363) by [Justin Gordon](https://github.com/justin808).
 - **Fixed `cpflow` crashing at load time with `invalid byte sequence in US-ASCII (ArgumentError)` on systems without a UTF-8 locale.** [PR 404](https://github.com/shakacode/control-plane-flow/pull/404) by [Justin Gordon](https://github.com/justin808). `Command::Base.all_commands` now reads command files with an explicit UTF-8 encoding instead of relying on `Encoding.default_external`. Fixes [issue 372](https://github.com/shakacode/control-plane-flow/issues/372).
 
+### Fixed
```
[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md]: Fixed in commit fd65819. The PR #395 entry now appends to the existing 5.2.0 Fixed section, and a structure check confirms the section has exactly one Fixed heading.

File: .github/workflows/cpflow-deploy-review-app.yml [hunk 15]
```diff
@@ -463,15 +463,79 @@ jobs:
...
+                  app_url = app_domain_template
+                    .gsub(cpln_gvc_alias_token, cpln_gvc_alias)
+                    .gsub("{{APP_NAME}}", app_name)
+
+                  begin
+                    app_uri = URI(app_url)
+                    if app_uri.path.to_s.empty? && app_uri.query.nil? && app_uri.fragment.nil?
+                      app_url = "#{app_url}/"
+                    end
+                  rescue URI::InvalidURIError
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/cpflow-deploy-review-app.yml:517]: [auto-deferred]

This is a minor robustness suggestion after the final candidate. Current malformed-input cases reach URI::InvalidURIError and are regression-tested; broadening the rescue now would spend another full review cycle and could mask a future programming defect. No correctness failure was identified.

File: spec/command/generate_github_actions_spec.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (justin808) [reply spec/command/generate_github_actions_spec.rb:67]: [auto-deferred]

This repeats the already-triaged test-helper indentation nit. The extractor is currently exercised by 68 passing examples, and changing a failure-message/formatting concern after the final candidate would restart CI and review without altering product behavior.

**Comment by justin808** [comment]
## Address-review summary

Scan scope: full PR review history through 2026-07-11T02:03:00Z.

### Mattered

- Fixed the empty-alias `nil`/`gsub` crash and added regression coverage in `61d2840`.
- Merged current `main` and refreshed generated command docs for 5.2.0 in `6780910`/`7a191f6`, clearing the hosted Command Docs failure.
- Removed the duplicate 5.2.0 `### Fixed` heading in `fd65819`.
- All affected review threads were replied to and resolved.

### Optional

- Documentation expansion, a bash-level mocking harness, repeated host lookup cleanup, output-file defense in depth, extractor formatting/failure-message cleanup, and broader URI rescue were auto-deferred or declined with thread-level rationale after the final-candidate debounce.

### Skipped

- None.

Deferred-work tracking: no review-feedback issue created; the required semantic-workflow post-merge exercise is tracked separately in https://github.com/shakacode/control-plane-flow/issues/407.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

**Comment by justin808** [comment]
### QA Evidence

- QA lane: `cpf-pr395-qa-2vl3ja`, `/private/tmp/control-plane-flow-pr395-qa`; ownership verified, final heartbeat `done`, claim released before coordinator closeout.
- Scope checked: exact-head changelog-only delta plus carried-forward independent review-app workflow regression evidence from the previously satisfied integrated head.
- Tested at: `fd658195cc3edee7b7c1e85516cefda76945118e`.
- Automated checks: focused workflow spec 68 examples/0 failures; RuboCop clean; generated command docs clean locally and hosted; YAML parse clean; changed workflow actionlint clean with the repository's known unsupported `job.workflow_*` diagnostics ignored; current-head RSpec Fast, RuboCop, Command Docs, Check Links, and `claude-review` pass.
- Manual checks: standard alias, org-prefixed alias, empty alias, custom endpoint, invalid URI, fallback wiring, workflow security-surface audit, and exact final changelog delta inspection.
- Findings: none; prior empty-alias, base-integration, and duplicate-heading blockers are fixed.
- QA required: yes.
- QA required rationale: semantic reusable-workflow behavior changed.
- QA lane status: satisfied.
- Release-blocking status: clear.
- Process-gap disposition: not applicable.
